src/workerd/jsg/ser-test.c++

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "jsg-test.h"
+#include "ser.h"
+
+namespace workerd::jsg::test {
+namespace {
+
+V8System v8System;
+class ContextGlobalObject: public jsg::Object, public ContextGlobal {};
+
+kj::Array<kj::byte> lastSerializedData;
+
+struct SerTestContext: public ContextGlobalObject {
+  enum class SerializationTag {
+    FOO,
+    BAR
+  };
+
+  struct Foo: public jsg::Object {
+    uint i;
+    Foo(uint i): i(i) {}
+
+    static jsg::Ref<Foo> constructor(uint i) {
+      return jsg::alloc<Foo>(i);
+    }
+
+    int getI() {
+      return i;
+    }
+
+    JSG_RESOURCE_TYPE(Foo) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(i, getI);
+    }
+
+    void serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+      serializer.writeRawUint32(i);
+    }
+    static jsg::Ref<Foo> deserialize(Lock& js, SerializationTag tag, Deserializer& deserializer) {
+      KJ_ASSERT(tag == SerializationTag::FOO);
+
+      // Intentionally deserialize differently so we can detect it.
+      return jsg::alloc<Foo>(deserializer.readRawUint32() + 2);
+    }
+    JSG_SERIALIZABLE(SerializationTag::FOO);
+  };
+
+  struct Bar: public jsg::Object {
+    kj::String text;
+    Bar(kj::String text): text(kj::mv(text)) {}
+
+    static jsg::Ref<Bar> constructor(kj::String text) {
+      return jsg::alloc<Bar>(kj::mv(text));
+    }
+
+    kj::String getText() {
+      return kj::str(text);
+    }
+
+    JSG_RESOURCE_TYPE(Bar) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(text, getText);
+    }
+
+    void serialize(jsg::Lock& js, jsg::Serializer& serializer) {
+      serializer.writeRawUint64(text.size());
```

Comment:
Is the assumption here that the serialize method always serializes to the most recent version `Tag` for this type?

Reply:
The first argument to `JSG_SERIALIZABLE(...)` is the tag that `serialize()` produces, yes.

Reply:
fwiw, I know this is covered a bit in the documentation in ser.h already, I just think it might need to be made more prominent. The note is a bit easy to miss.

Reply:
This should get a section in jsg/README.md once we've settled on the interface, but I suspect this first iteration won't actually be what we settle on, so I didn't want to over-document it just yet.

Having the serializer write its own tag would also eliminate the need for JSG_SERIALIZABLE_ONEWAY -- you'd just have JSG_SERIALIZABLE() instead. So maybe that is the best way. I'm just a little sad about the asymmetry -- the serializer writes its own tag but the parser doesn't read its own tag. But maybe that's fine.

I'd like to think about this a bit more before changing it, though. We can easily change this in a follow-up PR.